core/shared/src/main/scala/scalaz/zio/stream/Chunk.scala

```diff
@@ -0,0 +1,845 @@
+      else Chunk.Slice(Chunk.Arr(dest), 0, j)
+    }
+
+    override def dropWhile(f: A => Boolean): Chunk[A] = {
+      val self = array
+      val len  = self.length
+
+      var i = 0
+      while (f(self(i)) && i < len) {
+        i += 1
+      }
+
+      if (i == len) Chunk.empty
+      else Chunk.Slice(this, i, len - i)
+    }
+
+    override def filter(f: A => Boolean): Chunk[A] = {
+      val self = array
+      val len  = self.length
+      val dest = Array.ofDim[A](len)
+
+      var i = 0
+      var j = 0
+      while (i < len) {
+        val elem = self(i)
+
+        if (f(elem)) {
+          dest(j) = elem
+          j += 1
+        }
+
+        i += 1
+      }
+
+      if (dest == null) Chunk.Empty
+      else Chunk.Slice(Chunk.Arr(dest), 0, j)
+    }
+
+    override def flatMap[@specialized B](f: A => Chunk[B]): Chunk[B] = {
+      val self                   = array
+      val len                    = self.length
+      var chunks: List[Chunk[B]] = Nil
+
+      var i               = 0
+      var total           = 0
+      var B0: ClassTag[B] = null.asInstanceOf[ClassTag[B]]
+      while (i < len) {
+        val chunk = f(self(i))
+        if (chunk.length > 0) B0 = Chunk.classTagOf(chunk)
+        chunks ::= chunk
+        i += 1
+        total += chunk.length
+      }
+      chunks = chunks.reverse
+
+      if (B0 == null) Chunk.empty
+      else {
+        implicit val B = B0
+
+        val dest: Array[B] = Array.ofDim(total)
+
+        val it = chunks.iterator
+        var n  = 0
+        while (it.hasNext) {
+          val chunk = it.next
+
+          chunk.toArray(n, dest)
+          n += chunk.length
+        }
+
+        Arr(dest)
+      }
+    }
+
+    override def foldLeft[@specialized S](s0: S)(f: (S, A) => S): S = {
+      val self = array
+      val len  = self.length
+      var s    = s0
+
+      var i = 0
+      while (i < len) {
+        s = f(s, self(i))
+        i += 1
+      }
+
+      s
+    }
+
+    override def foldRight[@specialized S](s0: S)(f: (A, S) => S): S = {
+      val self = array
+      val len  = self.length
+      var s    = s0
+
+      var i = len - 1
+      while (i >= 0) {
+        s = f(self(i), s)
+        i -= 1
+      }
+
+      s
+    }
+
+    override def map[@specialized B](f: A => B): Chunk[B] = {
+      val self = array
+      val len  = self.length
+      var dest = null.asInstanceOf[Array[B]]
+
+      var i = 0
+      while (i < len) {
+        val b = f(self(i))
+
+        if (dest == null) {
+          implicit val B: ClassTag[B] = Chunk.Tags.fromValue(b)
+
+          dest = Array.ofDim[B](len)
+        }
+
+        dest(i) = b
+
+        i = i + 1
+      }
+
+      if (dest != null) Chunk.Arr(dest)
+      else Chunk.Empty
+    }
+
+    override def materialize[A1 >: A]: Chunk[A1] = this
+
+    override def takeWhile(f: A => Boolean): Chunk[A] = {
+      val self = array
+      val len  = self.length
+
+      var i = len - 1
+      while (!f(self(i)) && i >= 0) {
```

Comment:
Is this correct? The takeWhile always starts at the head and stops as soon as the condition is false. 

This tests fails:

`def takeWhile = Chunk(1, 2, 3, 4, 1).takeWhile(_ < 3) must_=== Chunk(1, 2)`

Reply:
Indeed